Implement argument checking for exported functions

ezraporter opened the pull request:
# Description
This PR implements error checking for arguments to exported functions with the following exceptions:
- `extract_tibbles()`: No checking for `tbls` since this takes an unevaluated expression and `eval_select()` does checking on our behalf with `cli`-generated messages
- `fmt_*` format helpers: No checking on arguments since these are all thin wrappers for `stringr` functions. `stringr` does a nice job of trying to coerce inputs to `character` and I don't think we want to override this behavior.

# Proposed Changes 

- Add internal check functions `check_arg_is_dataframe()`, `check_arg_is_env()`, `check_arg_is_character()`, `check_arg_is_logical()`,  `check_arg_choices()` that are wrappers for `checkmate` checks.
    - These wrappers are made with a function factory, `wrap_checkmate()`, so we can easily add new checking functions in the future
- Add tests to `test-checks.R` to make sure check functions work
- Add tests to test files for exported functions to check that they error with bad inputs

**Screenshots**
Example error message:

![Screen Shot 2022-12-15 at 11 46 57 AM](https://user-images.githubusercontent.com/60618324/207919174-9c6fc283-d1d8-4daf-94ea-ea2615e901e2.png)

The first line is added by `REDCapTidieR` and the second line is the message from `checkmate`.

### Issue Addressed
Closes #109

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [x] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

skadauke commented on R/read_redcap.R:
```diff
@@ -72,6 +72,14 @@ read_redcap <- function(redcap_uri,
                         forms = NULL,
                         export_survey_fields = TRUE,
                         suppress_redcapr_messages = TRUE) {
+
```
Please take a look at https://github.com/OuhscBbmc/REDCapR/blob/main/R/redcap-read.R to see how @wibeasley is checking `redcap_read()`'s parameters. I don't fully understand the need for some of the parameters but would be good to check we're not missing anything.

ezraporter replied:
- `any.missing = FALSE` disallows `NA`s. Makes sense to me to add this
- `pattern="^.{1,}$"` disallows empty strings (`""`). I could go either way here on needing this
- I see Will has `any.missing = TRUE` for `forms`. Not sure we want to accepts `forms` inputs with `NA`s

skadauke commented on R/read_redcap.R:
```diff
@@ -72,6 +72,14 @@ read_redcap <- function(redcap_uri,
                         suppress_redcapr_messages = TRUE) {
+
+  check_arg_is_character(redcap_uri, len = 1)
+  check_arg_is_character(token, len = 1)
```
For checking the token for validity, please take a look at https://github.com/OuhscBbmc/REDCapR/blob/main/R/sanitize-token.R. I think we might want to wrap this into its own check, e.g. `check_arg_is_valid_token()`

ezraporter replied:
It looks like the only additional thing `sanitize_token()` adds is the check that the token is length 32 hexadecimal but allows the user to override this check with the `REDCAP_TOKEN_PATTERN` environment variable. It seems like we'd need to duplicate that in a `check_arg_is_valid_token()`.

Are we sure we want to do that? It seems like duplication of custom code across packages that we'd want to always be in sync anyways.

skadauke replied:
I would, actually, because I think it should be the responsibility of the exported function to check args for validity. And I'm not talking about reimplementing `sanitize_token()` but actually calling it from our check function.

ezraporter replied:
Ah! I didn't realize `sanitize_token()` was exported. That's why I was confused

skadauke commented on R/read_redcap.R:
```diff
@@ -72,6 +72,14 @@ read_redcap <- function(redcap_uri,
+  check_arg_is_character(redcap_uri, len = 1)
+  check_arg_is_character(token, len = 1)
+  check_arg_choices(raw_or_label, choices = c("label", "raw"))
+  check_arg_is_character(forms, null.ok = TRUE)
```
OK for now, might change later if we decide to implement tidyselect grammar.

skadauke reviewed (changes requested):
This is an amazing start! My main comment is to tweak things to make it possible to create more informative error messages but I think we're close. 

At some point we should run the "bad input" tests manually to refine the exact error messages.

ezraporter commented on R/bind_tibbles.R:
```diff
@@ -43,6 +43,10 @@
 bind_tibbles <- function(supertbl,
                          environment = global_env(),
                          tbls = NULL) {
+  check_arg_is_dataframe(supertbl)
```
Sounds like a good idea. Not sure we want to check for presence of `redcap_metadata` here since `bind_tibbles()` doesn't need it anyways. Should `check_arg_is_supertbl()` possibly have options to specify which columns are required? Or is an eager check that always requires both columns fine?

rsh52 replied:
Is there any way that we can add `supertbl` as on object that gets returned from `class()`? Admittedly I don't know the details of what's required to make something like that happen.

rsh52 replied:
Since we're leaning so heavily on the supertibble being a specific thing unique to our package, it could be useful for validations like this, but I'm not sure if it's really the best use case for it.

http://adv-r.had.co.nz/OO-essentials.html#rc

ezraporter replied:
Yeah good idea. It looks like we could do it pretty easily with S3 (https://adv-r.hadley.nz/s3.html#s3-classes)

```r
# Would add this at the end of `read_redcap()`
class(supertbl) <- c("our_class", class(supertbl))

# Then check with
inherits(supertbl, "our_class")
```

My biggest question is whether we should do something like this in this PR or think about it more first. We can always implement a simple `check_arg_is_supertbl()` here that doesn't rely on S3 and then replace the internals later.

ezraporter replied:
Choosing to add a non-S3 `check_arg_is_supertbl()` for now but we can revisit

rsh52 replied:
Yea I think this is something we should implement outside of this PR, and if we want to move forward with it would make for a great addition to 0.3. Documentation could update to talking about how "not only is this a term we're using, but it's a reference-able class as well!"

skadauke replied:
Hm, I think adding a custom S3 class really isn't a big deal and would help identify incorrect inputs so think in scope for this PR. For the check, we might want to make sure that the class is correct and also that critical columns are there (`redcap_data`, `redcap_metadata`, what else?).

ezraporter replied:
One issue I'm running into:

In some of our documentation examples we make `supertbl` objects from scratch (see `bind_tibbles()` for example). If we want to keep those examples we'll need either:
1. An exported way to convert `tibbles` to `supertbls`
    - Not so hard to do but we wouldn't want users to actually use this function!
2. **Or** to manually set the S3 class in our examples
    - Works but really ugly documentation

Checks are currently failing because of the examples

ezraporter replied:
Actually came up with an **option 3** just after sending:

3. Export a pre-built example `supertbl` that we can use for the examples which already has the right class

I think this is actually the best solution. Thoughts?

rsh52 replied:
Option 3 LGTM 👍

ezraporter commented on R/checks.R:
```diff
@@ -272,3 +272,65 @@ check_req_labelled_metadata_fields <- function(supertbl) {
+#' @importFrom rlang caller_arg
+#' @importFrom cli cli_abort
+#' @noRd
+wrap_checkmate <- function(f, f_name = caller_arg(f)) {
```
Nope, moving this inside `wrap_checkmate()`!

skadauke commented on R/checks.R:
```diff
@@ -272,3 +272,65 @@ check_req_labelled_metadata_fields <- function(supertbl) {
+      return(TRUE)
+    }
+
+    cli_abort(
```
Found out how to do this here: https://rlang.r-lib.org/reference/topic-error-call.html

ezraporter commented on R/checks.R:
```diff
@@ -272,3 +272,65 @@ check_req_labelled_metadata_fields <- function(supertbl) {
+
+    cli_abort(
+      message = c(
+        "x" = "{.arg {arg}} is invalid",
```
> I'm wondering if the `"x"` and `"i"` part of the message should be optional arguments to the `wrap_checkmate()` function that can override the default. (Would have `NULL` as default for `"i"` but can supply more info if we want.)

To clarify, is the suggestion that we keep the `checkmate` message in `"!"` in all cases but add the ability to tack on additional lines of info to specific error messages?

skadauke replied:
What I was thinking is that the function definition of `wrap_checkmate()` could be:

```r
wrap_checkmate <- function(f, 
                           "x" = "You've supplied {y} for {.arg {arg}} which is not a valid value.",
                           "!" = "{out}",
                           "i" = NULL)
```

The function definition inside the function factory could pass these defaults on, allowing for modification (note, have to change `x` to `y` below to avoid conflict with `x` for the cli message.

```
  function(y, "x" = x, "!" = `!`, "i" = i, ..., arg = caller_arg(x)) {
```

And then you'd pass these values down to `cli_abort()`:

```r
    cli_abort(
      message = c(
        "x" = x,
        "!" = `!`,
        "i" = i
      ),
```

ezraporter commented:
Made the following updates:

- Added `call` to all `cli_*()` so the correct function name is shown in error messages
- Added `check_arg_is_supertbl()` with option to specify required columns
- Updated default `check_arg_*()` message to be more verbose
- Added capability to tack additional error message info onto error messages with `info` parameter in `check_arg_*()` functions
    - See below for examples of the error messages for various bad inputs to our exported function. **Looking for feedback on where we might need to add additional info.**
- Updated `checkmate` parameters based on settings in `REDCapR` to make argument checking a bit stricter

<details>
<summary>Error Message Examples</summary>

``` r
devtools::load_all()
#> ℹ Loading REDCapTidieR

options(rlang_backtrace_on_error_report = "none")

# read_redcap

classic_token <- "123456789ABCDEF"
redcap_uri <- "www.google.com"

## redcap_uri

read_redcap(123, classic_token)
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `redcap_uri`
#> ! Must be of type 'character', not 'double'

read_redcap(letters[1:3], classic_token)
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `redcap_uri`
#> ! Must have length 1, but has length 3

## token

read_redcap(redcap_uri, 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `token`
#> ! Must be of type 'character', not 'double'

read_redcap(redcap_uri, letters[1:3])
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `token`
#> ! Must have length 1, but has length 3

## raw_or_label

read_redcap(redcap_uri, classic_token, raw_or_label = "bad option")
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `raw_or_label`
#> ! Must be element of set {'label','raw'}, but is 'bad option'

## forms

read_redcap(redcap_uri, classic_token, forms = 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `forms`
#> ! Must be of type 'character' (or 'NULL'), not 'double'

## export_survey_fields

read_redcap(redcap_uri, classic_token, export_survey_fields = 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `export_survey_fields`
#> ! Must be of type 'logical', not 'double'

read_redcap(redcap_uri, classic_token, export_survey_fields = c(TRUE, TRUE))
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `export_survey_fields`
#> ! Must have length 1, but has length 2

## suppress_redcapr_messages

read_redcap(redcap_uri, classic_token, suppress_redcapr_messages = 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `suppress_redcapr_messages`
#> ! Must be of type 'logical', not 'double'

read_redcap(redcap_uri, classic_token, suppress_redcapr_messages = c(TRUE, TRUE))
#> Error in `read_redcap()`:
#> ✖ You've supplied an invalid value to `suppress_redcapr_messages`
#> ! Must have length 1, but has length 2

# bind_tibbles

bind_tibbles(123)
#> Error in `bind_tibbles()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> ! Must be of type 'data.frame', not 'double'

supertbl <- tibble(redcap_data = list())
bind_tibbles(supertbl, environment = "abc")
#> Error in `bind_tibbles()`:
#> ✖ You've supplied an invalid value to `environment`
#> ! Must be an environment, not 'character'

bind_tibbles(supertbl, tbls = 123)
#> Error in `bind_tibbles()`:
#> ✖ You've supplied an invalid value to `tbls`
#> ! Must be of type 'character' (or 'NULL'), not 'double'

# extract_tibbles

extract_tibbles(123)
#> Error in `extract_tibbles()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> ! Must be of type 'data.frame', not 'double'

# extract_tibble

extract_tibble(123, "my_tibble")
#> Error in `extract_tibble()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> ! Must be of type 'data.frame', not 'double'

supertbl <- tibble(redcap_data = list())
extract_tibble(supertbl, tbl = 123)
#> Error in `extract_tibble()`:
#> ✖ You've supplied an invalid value to `tbl`
#> ! Must be of type 'character', not 'double'

extract_tibble(supertbl, tbl = letters[1:3])
#> Error in `extract_tibble()`:
#> ✖ You've supplied an invalid value to `tbl`
#> ! Must have length 1, but has length 3

# make_labelled

make_labelled(123)
#> Error in `make_labelled()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> ! Must be of type 'data.frame', not 'double'

missing_col_supertbl <- tibble(redcap_data = list())
make_labelled(missing_col_supertbl)
#> Error in `make_labelled()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> ! Must contain columns `redcap_data` and `redcap_metadata`
#> ! You are missing `redcap_metadata`

missing_list_col_supertbl <- tibble(redcap_data = list(), redcap_metadata = 123)
make_labelled(missing_list_col_supertbl)
#> Error in `make_labelled()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> ! `redcap_metadata` must be of type 'list'
```

<sup>Created on 2022-12-16 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
</details>

rsh52 commented:
I think these returns largely look great. I especially like "element set" message for the `raw_or_label` return.

One thing to note, which may not be implementable here but for the future, is if `extract_tibble` winds up taking multiple arguments per Stephan's proposed framework in #111 we would need to allow for multiple `tbl`s (actually it probably needs to be updated to be `tbls` if we go forward with it).

For `make_labelled`, if and when we make a custom S3 class, it will be nice to provide an error saying something along the lines of "supertbl must be of class 'supertbl'" or something. The current one specifies just that certain colnames are missing:

```r
> make_labelled(data.frame(mtcars))
Error in `check_req_labelled_fields()` at REDCapTidieR/R/labelled.R:58:2:
! `supertbl` must contain `redcap_data` and `redcap_metadata`
✖ You are missing `redcap_data` and `redcap_metadata`
Run `rlang::last_error()` to see where the error occurred.
```

For fun I modified mtcars to have these colnames, and (while I this should never in a million years happen) here's what could happen:

```r
mtcars_mod <- data.frame(mtcars) |> 
  mutate(
    redcap_data = cyl,
    redcap_metadata = hp
  )

> make_labelled(data.frame(mtcars_mod))
Error in `check_req_labelled_metadata_fields()` at REDCapTidieR/R/labelled.R:59:2:
! All elements of `supertbl$redcap_metadata` must contain
  `field_name` and `field_label`
✖ `supertbl$redcap_metadata[[1]]` is missing `field_name` and
  `field_label`
✖ `supertbl$redcap_metadata[[2]]` is missing `field_name` and
  `field_label`
....
```

skadauke commented:
This looks great! A few comments:

- The message "You've supplied an invalid value to `xxx`" should additionally tell the user what that invalid value was unless I guess it's too complex.
- I think we can make the supertbl checks a bit nicer. I also like the idea of creating an S3 class for this, but I would also check for the presence of specific columns.

How about:

```
missing_col_supertbl <- tibble(redcap_data = list())
make_labelled(missing_col_supertbl)
#> Error in `make_labelled()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> i `supertbl` must be a REDCapTidieR supertibble, generated using `read_redcap()`
#> ! Must be of class `<redcaptidier_supertbl>`

missing_list_col_supertbl <- tibble(redcap_data = list(), redcap_metadata = 123)
make_labelled(missing_list_col_supertbl)
#> Error in `make_labelled()`:
#> ✖ You've supplied an invalid value to `supertbl`
#> i `supertbl` must be a REDCapTidieR supertibble, generated using `read_redcap()`
#> ! `redcap_metadata` must be of type 'list'
```

- We should handle the situation when someone uses `make_labelled()` but doesn't have labelled installed. I think this should throw an error and suggest running `install.packages("labelled")`.

ezraporter commented:
> * The message "You've supplied an invalid value to `xxx`" should additionally tell the user what that invalid value was unless I guess it's too complex.

I've been messing around with this but haven't quite gotten it.

If we want the actual evaluated value passed by the user it's easy:
```r
"You've supplied {x} for {.arg {arg}} which is not a valid value."
```

But this won't work for things like `check_arg_is_supertbl()` where we don't want to print out the actual `tibble()`.

If we want the actual unevaluated text the user entered it doesn't seem like there's an easy way to do this without capturing the value in the function that calls the `check_arg_*()` function.

<details>
<summary>Works but ugly</summary>

```r
library(rlang)

exported_function <- function(param) {
  check_arg_function(param, val = caller_arg(param))
}

check_arg_function <- function(x, val, arg = caller_arg(x)) {
  cat("The value of", arg, "was", val)
}

x <- 123

exported_function(x)
#> The value of param was x
exported_function(123)
#> The value of param was 123
```

</details>


> * I think we can make the supertbl checks a bit nicer. I like the idea of creating an S3 class for this, but I would also check for the presence of critical columns (see my comment above).
> 
> How about:
> 
> ```
> missing_col_supertbl <- tibble(redcap_data = list())
> make_labelled(missing_col_supertbl)
> #> Error in `make_labelled()`:
> #> ✖ You've supplied an invalid value to `supertbl`
> #> ! Must be of class `<redcaptidier_supertbl>`
> #> i `supertbl` must be a REDCapTidieR supertibble, generated using `read_redcap()`
> 
> missing_list_col_supertbl <- tibble(redcap_data = list(), redcap_metadata = 123)
> make_labelled(missing_list_col_supertbl)
> #> Error in `make_labelled()`:
> #> ✖ You've supplied an invalid value to `supertbl`
> #> ! `supertbl$redcap_metadata` must be of type 'list'
> #> i `supertbl` must be a REDCapTidieR supertibble, generated using `read_redcap()`
> ```

If we're good to add the S3 class I think this will all work in a modified version of `check_arg_is_supertbl()`

> * We should handle the situation when someone uses `make_labelled()` but doesn't have labelled installed. I think this should throw an error and suggest running `install.packages("labelled")`.

`check_installed()` currently handles this:

https://github.com/CHOP-CGTInformatics/REDCapTidieR/blob/7fc296b0b12b0b0ca5ecb532002ec7af3b63911f/R/labelled.R#L53

Any reason to change it?

skadauke commented:
> If we want the actual evaluated value passed by the user it's easy:
> 
> ```r
> "You've supplied {x} for {.arg {arg}} which is not a valid value."
> ```
> 
> But this won't work for things like `check_arg_is_supertbl()` where we don't want to print out the actual `tibble()`.
> 
> If we want the actual unevaluated text the user entered it doesn't seem like there's an easy way to do this without capturing the value in the function that calls the `check_arg_*()` function.

Got it. I think we could handle this like so:

- If {x} is an atomic vector <= 5 elements, show the whole thing
- If {x} is an atomic vector >5 element, truncate it at 5 (example here: https://style.tidyverse.org/error-messages.html#error-details). 
- If {x} is any other data structure, don't show the value, and instead use the message "You've supplied an invalid value for {.arg {arg}}."

ezraporter commented:
To summarize the additional pieces of this PR in the works today:
- Add `check_arg_is_valid_token()` with `sanitize_token()` from `REDCapR`
- Add S3 class to supertibble and update `check_arg_is_supertbl()` error messages as outlined [here](https://github.com/CHOP-CGTInformatics/REDCapTidieR/pull/114#issuecomment-1355686475)
- Update `wrap_checkmate()` error messages to include argument values as outlined [here](https://github.com/CHOP-CGTInformatics/REDCapTidieR/pull/114#issuecomment-1356292075)
- Update ability to customize `wrap_checkmate()` error messages (see [here](https://github.com/CHOP-CGTInformatics/REDCapTidieR/pull/114#discussion_r1051198924))
    - @skadauke I'm wondering if we can move this to a separate issue and the functionality if/when we need it. Currently I don't think we'll need this level of customization to produce any of the messages we want

skadauke commented:
All of this sounds good to me! Agree, let’s polish off this PR and see how it goes.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: ezraporter ***@***.***>
Sent: Monday, December 19, 2022 11:33:27 AM
To: CHOP-CGTInformatics/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Mention ***@***.***>
Subject: [External]Re: [CHOP-CGTInformatics/REDCapTidieR] Implement argument checking for exported functions (PR #114)


To summarize the additional pieces of this PR in the works today:

  *   Add check_arg_is_valid_token() with sanitize_token() from REDCapR
  *   Add S3 class to supertibble and update check_arg_is_supertbl() error messages as outlined here<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F114%23issuecomment-1355686475&data=05%7C01%7Ckadaukes%40chop.edu%7C9bde5a788d414d9d8b4208dae1dec3e0%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638070644154473770%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=UW7RfSAcMEigibg61LaMfxqg17UVXSXgMmsZ%2BfraGfs%3D&reserved=0>
  *   Update wrap_checkmate() error messages to include argument values as outlined here<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F114%23issuecomment-1356292075&data=05%7C01%7Ckadaukes%40chop.edu%7C9bde5a788d414d9d8b4208dae1dec3e0%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638070644154473770%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Iz2dWP1oBFd22%2FDQdk4k0WJmlEGUXVO%2F69JhU4YyL0A%3D&reserved=0>
  *   Update ability to customize wrap_checkmate() error messages (see here<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F114%23discussion_r1051198924&data=05%7C01%7Ckadaukes%40chop.edu%7C9bde5a788d414d9d8b4208dae1dec3e0%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638070644154473770%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=PG%2Bazj20mm%2Fvumpz5zP97pi3D6Mc8r1g4mmUfopkJOM%3D&reserved=0>)
     *   @skadauke<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fskadauke&data=05%7C01%7Ckadaukes%40chop.edu%7C9bde5a788d414d9d8b4208dae1dec3e0%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638070644154473770%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=baVE%2BSHprB72Erj3TWwSCiaefNhdAVU%2FXkhBly%2Fu7Nw%3D&reserved=0> I'm wondering if we can move this to a separate issue and the functionality if/when we need it. Currently I don't think we'll need this level of customization to produce any of the messages we want

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F114%23issuecomment-1357929938&data=05%7C01%7Ckadaukes%40chop.edu%7C9bde5a788d414d9d8b4208dae1dec3e0%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638070644154473770%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=MLa7foKVjiO%2Fzj%2BLGke0Fd%2BqO3UTcSWAjr2eXxcHx08%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWVNYAORJXHBZA5ZVC3WOCE5PANCNFSM6AAAAAAS76LERI&data=05%7C01%7Ckadaukes%40chop.edu%7C9bde5a788d414d9d8b4208dae1dec3e0%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638070644154473770%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=%2FSb8xEObOLsmpBh4qu2d7J1GBNDooLIsYtIvqiGczzk%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.

ezraporter commented:
> * Add `check_arg_is_valid_token()` with `sanitize_token()` from `REDCapR`

Done

> * Add S3 class to supertibble and update `check_arg_is_supertbl()` error messages as outlined [here](https://github.com/CHOP-CGTInformatics/REDCapTidieR/pull/114#issuecomment-1355686475)

Done with the addition of adding an example `superheroes_supertbl` to reference in examples

> * Update `wrap_checkmate()` error messages to include argument values as outlined [here](https://github.com/CHOP-CGTInformatics/REDCapTidieR/pull/114#issuecomment-1356292075)

Done. Using `cli::cli_vec()` to format `x` values for atomic vectors so they get truncated to 5 elements when printed. For non-atomic values I'm using `rlang::as_label()` to print a label for the value instead of the value itself. This seems to work pretty well. For example, environments become "`<env>`" and tibbles becomes `<tibble[rows,cols]>` with numbers of rows and columns shows. See the examples below.

> * Update ability to customize `wrap_checkmate()` error messages (see [here](https://github.com/CHOP-CGTInformatics/REDCapTidieR/pull/114#discussion_r1051198924))

Agreed to hold off on this

<details>
<summary>Error Message Examples</summary>

``` r
devtools::load_all()
#> ℹ Loading REDCapTidieR

options(rlang_backtrace_on_error_report = "none")

# read_redcap

classic_token <- "123456789ABCDEF123456789ABCDEF01"
redcap_uri <- "www.google.com"

## redcap_uri

read_redcap(123, classic_token)
#> Error in `read_redcap()`:
#> ✖ You've supplied `123` for `redcap_uri` which is not a valid value
#> ! Must be of type 'character', not 'double'

read_redcap(letters[1:3], classic_token)
#> Error in `read_redcap()`:
#> ✖ You've supplied `a`, `b`, `c` for `redcap_uri` which is not a valid
#>   value
#> ! Must have length 1, but has length 3

## token

read_redcap(redcap_uri, 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied `123` for `token` which is not a valid value
#> ! Must be of type 'character', not 'double'

read_redcap(redcap_uri, letters[1:3])
#> Error in `read_redcap()`:
#> ✖ You've supplied `a`, `b`, `c` for `token` which is not a valid value
#> ! Must have length 1, but has length 3

## raw_or_label

read_redcap(redcap_uri, classic_token, raw_or_label = "bad option")
#> Error in `read_redcap()`:
#> ✖ You've supplied `bad option` for `raw_or_label` which is not a valid
#>   value
#> ! Must be element of set {'label','raw'}, but is 'bad option'

## forms

read_redcap(redcap_uri, classic_token, forms = 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied `123` for `forms` which is not a valid value
#> ! Must be of type 'character' (or 'NULL'), not 'double'

## export_survey_fields

read_redcap(redcap_uri, classic_token, export_survey_fields = 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied `123` for `export_survey_fields` which is not a valid
#>   value
#> ! Must be of type 'logical', not 'double'

read_redcap(redcap_uri, classic_token, export_survey_fields = c(TRUE, TRUE))
#> Error in `read_redcap()`:
#> ✖ You've supplied `TRUE`, `TRUE` for `export_survey_fields` which is not
#>   a valid value
#> ! Must have length 1, but has length 2

## suppress_redcapr_messages

read_redcap(redcap_uri, classic_token, suppress_redcapr_messages = 123)
#> Error in `read_redcap()`:
#> ✖ You've supplied `123` for `suppress_redcapr_messages` which is not a
#>   valid value
#> ! Must be of type 'logical', not 'double'

read_redcap(redcap_uri, classic_token, suppress_redcapr_messages = c(TRUE, TRUE))
#> Error in `read_redcap()`:
#> ✖ You've supplied `TRUE`, `TRUE` for `suppress_redcapr_messages` which
#>   is not a valid value
#> ! Must have length 1, but has length 2

# bind_tibbles

bind_tibbles(123)
#> Error in `bind_tibbles()`:
#> ✖ You've supplied `123` for `supertbl` which is not a valid value
#> ! Must be of class <redcaptidier_supertbl>
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`

supertbl <- tibble(redcap_data = list())
bind_tibbles(supertbl, environment = "abc")
#> Error in `bind_tibbles()`:
#> ✖ You've supplied `<tibble[,1]>` for `supertbl` which is not a valid
#>   value
#> ! Must be of class <redcaptidier_supertbl>
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`

bind_tibbles(supertbl, tbls = 123)
#> Error in `bind_tibbles()`:
#> ✖ You've supplied `<tibble[,1]>` for `supertbl` which is not a valid
#>   value
#> ! Must be of class <redcaptidier_supertbl>
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`

# extract_tibbles

extract_tibbles(letters[1:10])
#> Error in `extract_tibbles()`:
#> ✖ You've supplied `a`, `b`, `c`, …, `i`, `j` for `supertbl` which is not
#>   a valid value
#> ! Must be of class <redcaptidier_supertbl>
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`

# extract_tibble

extract_tibble(123, "my_tibble")
#> Error in `extract_tibble()`:
#> ✖ You've supplied `123` for `supertbl` which is not a valid value
#> ! Must be of class <redcaptidier_supertbl>
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`

supertbl <- tibble(redcap_data = list()) %>%
  as_supertbl()
extract_tibble(supertbl, tbl = 123)
#> Error in `extract_tibble()`:
#> ✖ You've supplied `123` for `tbl` which is not a valid value
#> ! Must be of type 'character', not 'double'

extract_tibble(supertbl, tbl = letters[1:3])
#> Error in `extract_tibble()`:
#> ✖ You've supplied `a`, `b`, `c` for `tbl` which is not a valid value
#> ! Must have length 1, but has length 3

# make_labelled

make_labelled(123)
#> Error in `make_labelled()`:
#> ✖ You've supplied `123` for `supertbl` which is not a valid value
#> ! Must be of class <redcaptidier_supertbl>
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`

missing_col_supertbl <- tibble(redcap_data = list()) %>%
  as_supertbl()
make_labelled(missing_col_supertbl)
#> Error in `make_labelled()`:
#> ✖ You've supplied `<redcaptidier_supertbl[,1]>` for `supertbl` which is
#>   not a valid value
#> ! Must contain `supertbl$redcap_metadata`
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`

missing_list_col_supertbl <- tibble(redcap_data = list(), redcap_metadata = 123) %>%
  as_supertbl()
make_labelled(missing_list_col_supertbl)
#> Error in `make_labelled()`:
#> ✖ You've supplied `<redcaptidier_supertbl[,2]>` for `supertbl` which is
#>   not a valid value
#> ! `supertbl$redcap_metadata` must be of type 'list'
#> ℹ `supertbl` must be a REDCapTidieR supertibble, generated using
#>   `read_redcap()`
```

<sup>Created on 2022-12-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
</details>

ezraporter commented:
@skadauke @rsh52 This is ready to review once checks pass

rsh52 commented on R/data.R:
```diff
@@ -0,0 +1,32 @@
+#'   \item{form_status_complete}{REDCap instrument completed?}
+#' }
+#' @source <https://www.superherodb.com/>
+"superheroes_supertbl"
```
Is this something we want to have exported to users? I believe as it is right now this will be available as `REDCapTidieR::superheroes_supertbl`, but is it meant to be internal?

ezraporter replied:
Maybe I'm wrong but I think this needs to be external external so we can use it in examples. I guess alternatively we could reference it in the examples like:

```r
superheroes_supertbl <- readRDS(system.file("testdata/superheroes_supertbl.RDS", package = "REDCapTidieR"))
```

but it won't look the best in our docs. Are there other options I'm missing?

rsh52 replied:
Ah, I didn't realize this would be a part of "option 3", my fault entirely. I think it's fine, we rely on it so heavily for our example documentation and you have the help documentation available to support it.

rsh52 commented on data-raw/superheroes_supertbl.R:
```diff
@@ -0,0 +1,6 @@
+
+superheroes_supertbl <- read_redcap(redcap_uri, token)
+
+usethis::use_data(superheroes_supertbl, overwrite = TRUE)
```
Just to keep track of the new folders, so we're introducing a `data-raw` folder (in `.Rbuildignore`) and a `data` folder (not in `.Rbuildignore`) that hold a script and a datafile, respectively.

May be something to do separately, but in the past I had put testing data under `/inst/testdata` (also in `.Rbuildignore`, with a `create_test_data.R` script) so we should consolidate these under your new folders.or under `/inst`.

ezraporter replied:
I think related to my comment above. If this was only internal testing data `/inst/testdata` would make sense.

I was following [this section](https://r-pkgs.org/data.html#sec-data-data) on adding exported data to a package and generated those new directories with `usethis::use_data_raw()` and `usethis::use_data()` described in there but if we have a way besides exporting this we can roll those directories back.

rsh52 replied:
If it's a process used in R packages and a part of the `usethis` framework then ignore me haha.

rsh52 reviewed (commented):
Just a couple of comments, once we make a decision on them everything else LGTM
